pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,439 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supportbundlecollection
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/sftp"
+	"github.com/spf13/afero"
+	"golang.org/x/crypto/ssh"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/exec"
+
+	"antrea.io/antrea/pkg/agent"
+	"antrea.io/antrea/pkg/agent/config"
+	agentquerier "antrea.io/antrea/pkg/agent/querier"
+	"antrea.io/antrea/pkg/apis/controlplane"
+	cpv1b2 "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/ovs/ovsctl"
+	"antrea.io/antrea/pkg/querier"
+	"antrea.io/antrea/pkg/support"
+	"antrea.io/antrea/pkg/util/compress"
+)
+
+type ProtocolType string
+
+const (
+	sftpProtocol ProtocolType = "sftp"
+
+	controllerName = "SupportBundleCollectionController"
+
+	uploadToFileServerTries      = 5
+	uploadToFileServerRetryDelay = 5 * time.Second
+)
+
+var (
+	emptyWatch      = watch.NewEmptyWatch()
+	defaultFS       = afero.NewOsFs()
+	defaultExecutor = exec.New()
+	// Declared as variable for testing.
+	newAgentDumper = support.NewAgentDumper
+
+	getUploader          = getUploaderByProtocol
+	sftpUploaderInstance *sftpUploader
+)
+
+type SupportBundleController struct {
+	nodeName                      string
+	nodeType                      config.NodeType
+	namespace                     string
+	antreaClientGetter            agent.AntreaClientProvider
+	queue                         workqueue.Interface
+	supportBundleCollections      map[string]*cpv1b2.SupportBundleCollection
```

Comment:
Is it working if we only maintain an object with type `*cpv1b2.SupportBundleCollection` instead of a map?

Reply:
Since controller side can guarantee that only send one support bundle collection to the agent at one time, I have changed it to a single object. Please help review the latest version.